### PR TITLE
REPORT-120:Create TX_TB_PREV Report

### PR DESCRIPTION
**Pull request by fghhisrepo**
The TB_PREV indicator reports the proportion of ART patients who started on a standard course of TB Preventive Treatment (TPT) in the previous reporting period who completed therapy.

### CODE REVIEW

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java [hunk 1]
```diff
@@ -0,0 +1,179 @@
...
+            + "union "
+            + "select p.patient_id,max(e.encounter_datetime) data_transferidopara from patient p "
+            + "inner join encounter e on p.patient_id=e.patient_id "
+            + "inner join obs o on o.encounter_id=e.encounter_id where 	e.voided=0 and p.voided=0 and e.encounter_datetime<=:endDate and "
+            + "o.voided=0 and o.concept_id=2016 and o.value_coded in (1706,23863) and e.encounter_type=21 and  e.location_id=:location group by p.patient_id "
+            + "union "
+            + "select p.patient_id,max(o.obs_datetime) data_transferidopara from patient p "
+            + "inner join encounter e on p.patient_id=e.patient_id "
+            + "inner join obs o on o.encounter_id=e.encounter_id "
+            + "where 	e.voided=0 and p.voided=0 and o.obs_datetime<=:endDate and "
```
Comment (Guimino) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: @emaposse, the same for **Ficha Resumo**

   > ps.start_date>=:**startDate – 6 months** and ps.start_date<=:endDate

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+            + "o.voided=0 and o.concept_id=6272 and o.value_coded=1706 and e.encounter_type=53 and  e.location_id=:location group by p.patient_id "
+            + "union "
+            + "select p.patient_id,max(e.encounter_datetime) data_transferidopara from	patient p "
+            + "inner join encounter e on p.patient_id=e.patient_id "
+            + "inner join obs o on o.encounter_id=e.encounter_id "
+            + "where e.voided=0 and p.voided=0 and e.encounter_datetime<=:endDate and "
```
Comment (Guimino) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: @emaposse, the same for **Ficha Clinica**

ps.start_date>=:**startDate – 6 months** and ps.start_date<=:endDate

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxTbPrevCohortQueries.java [hunk 3]
```diff
@@ -0,0 +1,193 @@
...
+
+    definition.setName("findPatientWhoStartedTpi");
+    definition.addParameter(new Parameter("startDate", "Start Date", Date.class));
+    definition.addParameter(new Parameter("endDate", "End Date", Date.class));
+    definition.addParameter(new Parameter("location", "location", Location.class));
+
+    definition.setQuery(TxTbPrevQueriesInterface.QUERY.findPatientsTransferredOut);
+
+    return definition;
+  }
```
Comment (Guimino) [inline api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxTbPrevCohortQueries.java]: @emaposse  I don't see the exclusions for patients whose after being marked as transferred out still pickingup or having  consultation.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/tbprev/TxTbPrevCalculation.java [hunk 4]
```diff
@@ -0,0 +1,7 @@
+package org.openmrs.module.eptsreports.reporting.calculation.tbprev;
+
+import org.openmrs.module.eptsreports.reporting.calculation.BaseFghCalculation;
+import org.springframework.stereotype.Component;
+
+@Component
+public class TxTbPrevCalculation extends BaseFghCalculation {}
```
[the remark this answers is not in the document]
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/calculation/tbprev/TxTbPrevCalculation.java]: Yes I will remove.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxTbPrevCohortQueries.java (hunk 3 above)
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxTbPrevCohortQueries.java]: @Guimino  this part of the code solves that problem
https://github.com/FriendsInGlobalHealth/openmrs-module-eptsreports/blob/e5b69e92ff2c5da382fd630aa9673e109d77284b/api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java#L173-L177

Comment (Guimino) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/cohorts/TxTbPrevCohortQueries.java]: Ok Cool

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java [hunk 5]
```diff
@@ -0,0 +1,179 @@
...
+            + "o.concept_id=6122 and o.value_coded in (1257,1065) and o.voided=0 and e.location_id=:location "
+            + "group by inicio_tpi.patient_id having count(*)>=6 ";
+
+    public static final String findPatientsTransferredOut =
+        "select transferidopara.patient_id from ( "
+            + "select patient_id,max(data_transferidopara) data_transferidopara from ( "
+            + "select pg.patient_id,max(ps.start_date) data_transferidopara from patient p "
+            + "inner join patient_program pg on p.patient_id=pg.patient_id "
+            + "inner join patient_state ps on pg.patient_program_id=ps.patient_program_id "
+            + "where pg.voided=0 and ps.voided=0 and p.voided=0 and  pg.program_id=2 and ps.state=7 and ps.end_date is null and  ps.start_date<=:endDate and location_id=:location group by p.patient_id "
```
[the remark this answers is not in the document]
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: done

[the remark this answers is not in the document]
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: Fixed @Guimino.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java [hunk 6]
```diff
@@ -0,0 +1,179 @@
...
+            + "union "
+            + "select p.patient_id,max(e.encounter_datetime) data_transferidopara from patient p "
+            + "inner join encounter e on p.patient_id=e.patient_id "
+            + "inner join obs o on o.encounter_id=e.encounter_id where 	e.voided=0 and p.voided=0 and e.encounter_datetime<=:endDate and "
```
[the remark this answers is not in the document]
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: Fixed @Guimino.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java (hunk 1 above)
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: Fixed @Guimino.

File: api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java (hunk 2 above)
Comment (emaposse) [reply api/src/main/java/org/openmrs/module/eptsreports/reporting/library/queries/TxTbPrevQueriesInterface.java]: Fixed @Guimino.

**Pull request merged**